### PR TITLE
Tool Submission: RoadFury  Transformer+SWA Test Prioritizer

### PR DESCRIPTION
**Pull request by chisngyen**
### Overview
RoadFury is a novel test prioritization tool for simulation-based regression testing of self-driving cars, developed for the ICST 2026 SDC Testing Competition.

### Approach
- **Transformer Encoder** (d=128, 4 layers, 8 heads) with learnable [CLS] token and Pre-LN architecture
- **10-channel road geometry features**: curvature, heading sin/cos, curvature jerk, curvature acceleration, segment length, angle change, cumulative distance, relative position, local curvature std
- **Stochastic Weight Averaging (SWA)**: averages weights over last 25 epochs for flatter minima and better generalization (Izmailov et al., 2018)
- **Focal Loss** (gamma=2.5): focuses learning on hard borderline examples
- **Training data**: 36,006 labeled tests from SensoDat (28,804 train / 7,202 val)

### Performance (Multi-trial Competition APFD)
| Tool | APFD |
|------|------|
| **RoadFury (Transformer+SWA+FocalLoss)** | **0.8066  0.0124** |
| ITEP4SDC (current SOTA) | 0.7812 |
| Random baseline | 0.4930 |

### Docker
\\\ash
docker build -t road-fury tools/prioritizers/road_fury
docker run --rm -t -p 4545:4545 road-fury -p 4545
\\\

### Files
| File | Description |
|------|-------------|
| main.py | gRPC server implementing Name, Initialize, Prioritize |
| features.py | 10-channel road geometry feature extraction |
| roadfury_best.pt | Pre-trained Transformer+SWA model (3.2 MB) |
| Dockerfile | Docker container configuration |
| LICENSE | MIT License |

### Author
Chi-Nguyen Tran (tcnguyen2365@gmail.com)

### CODE REVIEW

**Comment by ChristianBirchler** [comment]
Dear @chisngyen,

Thank you for your submission for the SDC Tool Competition at ICST 2026.
Your submission is well-received. We are happy to evaluate your tool.
For this, we might still reach out to you in case we encounter technical problems running the tool. From our past experience, this is unlikely to be the case.

In order to go further with the tool competition, we kindly ask you to send us an email the following information:
- Title and abstract of your tool paper
- Ordered list of authors, affiliations, and email

For the competition, you should prepare a two-page tool paper (including references) describing your tool and approach/algorithm.
Please use the same formatting guidelines as for the ICST research track.
We will do a brief review of your paper before it eventually gets published so that you can incorporate the feedback.
Therefore, share your paper with us by March 15 AoE by email (christian.birchler@unibe.ch).

Further information you can find on the website: https://conf.researchr.org/track/icst-2026/icst-2026-tool-competition--self-driving-car-testing

Again, thank you for taking part in the competition.

Best regards,
Christian
